### ISSUE TITLE
vm_compute hangs on a large expression

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#3222]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 3222
Date: 2014-01-27 16:29:45 +0100
From: Ken Roe &lt;<kendroe@hotmail.com>&gt;
To:   
CC:   coq-bugs-redist@lists.gforge.inria.fr, @_gares, @_silene, @_JasonGross

Last updated: 2015-03-29 18:15:14 +0200

## Bugzilla Comment ID: 6470
Date: 2014-01-27 16:29:45 +0100
From: Ken Roe &lt;<kendroe@hotmail.com>&gt;

I have attached an archive with two directories (which should unzip on the Mac), Coq4.vm_good and Coq4.vm_bad.  In the first, you can fully compile the file TreeTraversal.v.  In the second, the compilation will hang on the first call to the Lcac macro "simp".  "simp" is defined in the file Simplify.v.  The only change in the two models is that in Coq4.vm_good, "simp" uses a call to compute.  In Coq4.vm_bad, "simp" tries to use vm_compute.

                - Ken

## Bugzilla Comment ID: 6471
Date: 2014-01-27 16:40:43 +0100
From: Ken Roe &lt;<kendroe@hotmail.com>&gt;

Created attachment 436
Coq4.vm_good and Coq4.vm_bad

Unzip on the Mac and contains two Coq models.  You will need to execute "./compile" in each of the directories to compile all files except for TreeTraversal.v.

> Attached file: Archive.zip (application/download, 226810 bytes)
> Description:   Coq4.vm_good and Coq4.vm_bad

## Bugzilla Comment ID: 6472
Date: 2014-01-27 16:53:57 +0100
From: @_gares

I did not try to execute your code, but I see "eapply ...; vm_compute".
As fas as I know, vm_compute in 8.4 does not like the evars that eapply is likely to generate.  

I think the trunk version of Coq has been fixed w.r.t. that problem.  Would you mind running your test on Coq trunk?

Cheers

## Bugzilla Comment ID: 6474
Date: 2014-01-28 06:52:13 +0100
From: Ken Roe &lt;<kendroe@hotmail.com>&gt;

If you look carefully at the code in Simplify.v it is designed to only execute vm_compute on the portion of the expression that does not contain existentials.

Also, how do I download the trunk version?

## Bugzilla Comment ID: 6475
Date: 2014-01-28 07:57:58 +0100
From: @_JasonGross

You can download the trunk version of Coq at https://github.com/coq/coq/; either "download zip" (on the right) or clone it using git.

## Bugzilla Comment ID: 6476
Date: 2014-01-28 07:58:19 +0100
From: @_JasonGross

Note that you will have to compile the trunk version yourself.

## Bugzilla Comment ID: 6478
Date: 2014-01-28 18:14:09 +0100
From: Ken Roe &lt;<kendroe@hotmail.com>&gt;

Created attachment 437
Fixed version of compile script

Replace the version in the Coq4.vm_good and Coq4.vm_bad directories.  The change should not make any difference in testing the bug.  I just removed the line that was compiling SatSolver.v.  This file is not needed by TreeTraversal.v.

> Attached file: compile (application/octet-stream, 241 bytes)
> Description:   Fixed version of compile script

## Bugzilla Comment ID: 6479
Date: 2014-01-28 18:14:52 +0100
From: Ken Roe &lt;<kendroe@hotmail.com>&gt;

I just tested with Coq trunk.  I get the same result.  The bug is not fixed.

## Bugzilla Comment ID: 10066
Date: 2015-03-29 18:15:14 +0200
From: @_silene

I am unable to reproduce the bug on 8.5. And while there is indeed some sort of a hang on 8.4, it is unrelated to vm_compute.

Indeed, as far as I can tell, vm_compute terminates instantly in your testcase. Then you are calling "refine (@eq_refl X' <: X' = X')". So there is a second vm_compute involved there, but again it terminates instantly. What then takes ages is "refine" trying to match (X' = X') against (?42 = X) and that does not involve vm_compute. (By the way, this also means that your use of vm_compute is counter-productive, since you are asking Coq to compute more than it should.)

In other words, if there is a bug, it is in "refine" rather than in "vm_compute", and since "refine" was completely rewritten in 8.5, the bug now seems fixed.
